rs/messaging/tests/memory_tests.rs

```diff
@@ -0,0 +1,500 @@
+    const SHUTDOWN_PHASE_MAX_ROUNDS: u64 = 300;
+    // The amount of memory available for guaranteed response message memory on `local_env`.
+    const LOCAL_MESSAGE_MEMORY_CAPACITY: u64 = 100 * MB;
+    // The amount of memory available for guaranteed response message memory on `remote_env`.
+    const REMOTE_MESSAGE_MEMORY_CAPACITY: u64 = 50 * MB;
+
+    let fixture = Fixture::new(FixtureConfig {
+        local_canisters_count: 2,
+        local_max_instructions_per_round: 100_000_000,
+        local_message_memory_capacity: LOCAL_MESSAGE_MEMORY_CAPACITY,
+        remote_canisters_count: 1,
+        remote_max_instructions_per_round: 100_000_000,
+        remote_message_memory_capacity: REMOTE_MESSAGE_MEMORY_CAPACITY,
+    });
+
+    let config = CanisterConfig::try_new(
+        fixture.canisters(),   // receivers
+        0..=max_payload_bytes, // call_bytes
+        0..=max_payload_bytes, // reply_bytes
+        0..=0,                 // instructions_count
+    )
+    .unwrap();
+
+    // Send configs to canisters, seed the rng.
+    for (index, canister) in fixture.canisters().into_iter().enumerate() {
+        fixture.replace_config(canister, config.clone()).unwrap();
+        fixture.seed_rng(canister, seeds[index]);
+        fixture
+            .replace_reply_weight(canister, reply_weight)
+            .unwrap();
+        fixture.replace_call_weight(canister, call_weight).unwrap();
+    }
+
+    // Start chatter on all canisters.
+    fixture.start_chatter(calls_per_round).unwrap();
+
+    // Build up backlog and keep up chatter for while.
+    for _ in 0..CHATTER_PHASE_ROUND_COUNT {
+        fixture.tick();
+
+        // Check message memory limits are respected.
+        let (local_memory, remote_memory) = fixture.guaranteed_response_message_memory_taken();
+        if local_memory > LOCAL_MESSAGE_MEMORY_CAPACITY.into() {
+            return fixture.debug_info("Chatter: local memory exceeds limit");
+        }
+        if remote_memory > REMOTE_MESSAGE_MEMORY_CAPACITY.into() {
+            return fixture.debug_info("Chatter: remote memory exceeds limit");
+        }
+    }
+
+    // Stop chatter on all canisters; stop additional downstream calls.
+    fixture.stop_chatter().unwrap();
+    for canister in fixture.canisters() {
+        fixture.replace_call_weight(canister, 0).unwrap();
+    }
+
+    // Keep ticking until all calls are answered.
+    for counter in 0.. {
+        fixture.tick();
+
+        // Check message memory limits are respected.
+        let (local_memory, remote_memory) = fixture.guaranteed_response_message_memory_taken();
+        if local_memory > LOCAL_MESSAGE_MEMORY_CAPACITY.into() {
+            return fixture.debug_info("Shutdown: local memory exceeds limit");
+        }
+        if remote_memory > REMOTE_MESSAGE_MEMORY_CAPACITY.into() {
+            return fixture.debug_info("Shutdown: remote memory exceeds limit");
+        }
```

Comment:
You could extract this into a function. Maybe even the zero check below. You could make `guaranteed_response_message_memory_taken()` into `expect_guaranteed_response_message_memory_taken(limit: u64)`.

Reply:
Good suggestion, made a function.